### ISSUE TITLE
LDAP based console logins are very slow

### ISSUE THREAD

**Issue by dharapvj** [minio/minio#20214]
## Issue description
Similar to what was noted in #19746 and #19945, we see significantly slower login and UI display for LDAP logins only.

As suggested in those tickets, I have upgraded my tenant to latest release `RELEASE.2024-08-03T04-33-23Z` before submitting bug report.

I am attaching trace below as suggested:

```shell
# when logging in with a LDAP user
bash-5.1# mc admin trace -a --response-duration 250ms minio/
2024-08-05T04:32:28.818 [200 OK] sts.AssumeRoleWithSSO minio-test-ss-0-1.minio-test-hl.minio-test.svc.cluster.local:9000/ 10.244.9.51      631.227ms    ⇣  631.222093ms   ↑ 289 B ↓ 906 B
2024-08-05T04:32:30.095 [200 OK] admin.AccountInfo minio-test-ss-0-0.minio-test-hl.minio-test.svc.cluster.local:9000/minio/admin/v3/accountinfo 10.244.13.9      4.261281s    ⇣  4.261245171s   ↑ 259 B ↓ 44 KiB
2024-08-05T04:32:34.943 [200 OK] admin.AccountInfo minio-test-ss-0-0.minio-test-hl.minio-test.svc.cluster.local:9000/minio/admin/v3/accountinfo 10.244.9.51      4.729308s    ⇣  4.72929662s  ↑ 259 B ↓ 44 KiB
2024-08-05T04:32:44.093 [SCANNER] scanner.ScanCycle minio-test-ss-0-1.minio-test-hl.minio-test.svc.cluster.local:9000  389.834441ms

# when logging in with root user (non ldap) Also reduced time to 50ms to get something logged
bash-5.1# mc admin trace -a --response-duration 50ms minio/
2024-08-05T05:15:36.550 [400 Bad Request] sts.AssumeRoleWithSSO minio-test-ss-0-1.minio-test-hl.minio-test.svc.cluster.local:9000/ 10.244.13.9      171.355ms    ⇣  171.346591ms   ↑ 281 B ↓ 311 B
^Cbash-5.1# 

```
See corresponding network traces..

with LDAP based login - total time 12.47 seconds
![image](https://github.com/user-attachments/assets/0bfe9f34-bf41-4707-ab0d-34ab0580e8d2)

with root login - total time 1.49 seconds
![image](https://github.com/user-attachments/assets/a38cf896-a837-4271-8e2a-16a2693efedf)

#### With earlier version of minio
Minio version : `RELEASE.2024-03-15T01-07-19Z`
For LDAP user. Time taken - 5.78 seconds (is about 45% as compared to new version.)
![image](https://github.com/user-attachments/assets/89b69693-0674-48bd-adad-80e4aab6ae4c)

For Root user - time taken is almost the same. 1.29 seconds
![image](https://github.com/user-attachments/assets/c7766e76-8318-4f50-8f69-515113c3000f)

## Expected Behavior

Console access should remain speedy for both regular user as well as LDAP user

## Current Behavior

Console home page is rendered quickly only for non-ldap user

## Steps to Reproduce (for bugs)

1. Enable LDAP login for minio tenant via environment variables
2. Create at least 50 buckets in the tenant via `for i in {1..50}; do mc mb testbucket${i}; done`
3. Login to console using root access. Observe the speed of rendering home page (list all bucket)
4. Login to console using LDAP based user. Observe the MUCH SLOWER speed of rendering home page (list all bucket)

## Regression
 Yes. `RELEASE.2024-03-15T01-07-19Z` where we do not see this issue.

## Your Environment

* Version used (`minio --version`): `RELEASE.2024-08-03T04-33-23Z`
* Server setup and configuration: Minio operator based tenant
* Operating System and version (`uname -a`): ubuntu 20.04 (kubernetes worker node OS) minio is used in kubernetes cluster deployed via Minio operator

As a sidenote - latest version seems to be compiled with Go 1.22 and faces issues when connecting to LDAP
exactly same as noted [here](https://github.com/goharbor/harbor/issues/20691). And workaround fix as noted in [this comment on the same issue](https://github.com/goharbor/harbor/issues/20691#issuecomment-2220394365). It is not really a issue with minio product.. but in general go 1.22 and TLS compatibility issue. You might want add this as known issue workaround.

**Comment by harshavardhana** [comment]
> As a sidenote - latest version seems to be compiled with Go 1.22 and faces issues when connecting to LDAP
> exactly same as noted [here](https://github.com/goharbor/harbor/issues/20691). And workaround fix as noted in [this comment on the same issue](https://github.com/goharbor/harbor/issues/20691#issuecomment-2220394365). It is not really a issue with minio product.. but in general go 1.22 and TLS compatibility issue. You might want add this as known issue workaround.

Does this workaround work for you?

**Comment by dharapvj** [comment]
Yes it does.

सोम, ५, ऑग २०२४, २१:३० ला Harshavardhana ने ***@***.***>
लिहिला:

**Comment by harshavardhana** [comment]
So looks like this might be a problem not just for LDAP, it might be a problem for all TLS RSA ciphers in some sense.

**Comment by dharapvj** [comment]
The tls error is going to be an issue for many people using TLS RSA.
Already many threads about it on various GitHub projects.

But that issue is unrelated to main issue is slow LDAP based logics. Just
wanted to clarify that.

सोम, ५, ऑग २०२४, २१:५९ ला Harshavardhana ने ***@***.***>
लिहिला:

**Comment by harshavardhana** [comment]
> But that issue is unrelated to main issue is slow LDAP based logics. Just wanted to clarify that.

yeah that is because of timeouts due to the RSA key exchange problems.

**Comment by harshavardhana** [comment]
can you try this workaround? `MINIO_API_SECURE_CIPHERS=off` instead of the GODEBUG?

**Comment by dharapvj** [comment]
The main issue of slowness is suffered even in `RELEASE.2024-05-01T01-11-10Z` which does not have any issues in connecting to LDAP with TLS handshake error. The slowness has been present fro I think 1 release in April.

> can you try this workaround? MINIO_API_SECURE_CIPHERS=off instead of the GODEBUG?

I will try this and report.

**Comment by harshavardhana** [comment]
Does GODEBUG workaround work for minio.RELEASE.2024-05-01T01-11-10Z as well?

**Comment by dharapvj** [comment]
GODEBUG was not required in that release.

मंगळ, ६, ऑग २०२४, १३:१६ ला Harshavardhana ने ***@***.***>
लिहिला:

**Comment by harshavardhana** [comment]
> GODEBUG was not required in that release.

not sure I understand?

**Comment by dharapvj** [comment]
Sorry to not provide a clear information earlier

So.. With different releases I below difference in behavior..

`RELEASE.2024-03-26T22-10-45Z` - No slowness in Console access via LDAP
`RELEASE.2024-05-01T01-11-10Z` - Slowness can be seen for LDAP users. No tls handshake issues to connect to LDAP from minio.
`RELEASE.2024-08-03T04-33-23Z` - Now tls handshake issues appear when minio tries to connect to LDAP (and so we cannot login for any LDAP users). This TLS handshake error can be worked around via `GODEBUG` workaround. With the workaround, LDAP user can login but slowness can STILL be seen for LDAP users.

In summary - any release after April - you can see the LDAP user slowness (just create about 50 odd buckets in the minio instance as outlined in the original issue description).

**Comment by harshavardhana** [comment]
Was able to reproduce fix in https://github.com/minio/minio/pull/20320

**Comment by harshavardhana** [comment]
Closing as duplicate of https://github.com/minio/minio/issues/19746
